subprojects/core-api/src/main/java/org/gradle/api/artifacts/dsl/DependencyConstraintHandler.java

```diff
@@ -17,20 +17,74 @@
 
 import org.gradle.api.Action;
 import org.gradle.api.artifacts.DependencyConstraint;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderConvertible;
 
 /**
  * <p>A {@code DependencyConstraintHandler} is used to declare dependency constraints.</p>
  *
+ *
+ * <h2>Dependency constraint notations</h2>
+ *
+ * <p>There are several supported dependency constraint notations. These are described below. For each dependency constraint declared this
+ * way, a {@link DependencyConstraint} object is created. You can use this object to query or further configure the
+ * dependency constraint.</p>
+ *
+ * <p>You can also always add instances of {@link DependencyConstraint} directly:</p>
+ *
+ * <code><i>configurationName</i>(&lt;instance&gt;)</code>
+ *
+ * <p>Dependency constraints can also be declared with a {@link org.gradle.api.provider.Provider} that provides any of the other supported dependency constraint notations.</p>
+ *
+ * <h3>External dependencies</h3>
+ *
+ * <p>There are two notations supported for declaring a dependency constraint on an external module.
+ * One is a string notation formatted this way:</p>
+ *
+ * <code><i>configurationName</i>("<i>group</i>:<i>name</i>:<i>version</i>")</code>
+ *
+ * <p>The other is a map notation:</p>
+ *
+ * <code><i>configurationName</i>(group: <i>group</i>, name: <i>name</i>, version: <i>version</i>)</code>
+ *
+ * <p>In both notations, all properties, except name, are optional.</p>
+ *
+ * <pre class='autoTested'>
+ * plugins {
+ *     id 'java' // so that we can use 'implementation', 'testImplementation' for dependency constraints
+ * }
+ *
+ * dependencies {
+ *   constraints {
+ *     //for dependencies found in artifact repositories you can use
+ *     //the string notation, e.g. group:name:version
+ *     implementation 'commons-lang:commons-lang:2.6'
+ *     testImplementation 'org.mockito:mockito:1.9.0-rc1'
+ *
+ *     //map notation:
+ *     implementation group: 'com.google.code.guice', name: 'guice', version: '1.0'
+ *   }
+ * }
+ * </pre>
+ *
+ * <h3>Project dependencies</h3>
+ *
+ * <p>To add a project dependency constraint, you use the following notation:
+ * <p><code><i>configurationName</i>(project(":some-project"))</code>
+ *
+ * <p>The notation <code>project(":project-a")</code> is similar to the syntax you use
+ * when configuring a projectA in a multi-module gradle project.
+ *
```

Comment:
I'd leave this out (and maybe remove it from the `DependencyHandler` docs) since it isn't really Isolated Projects friendly